Add repair step for outdated OCS IDs

LukasReschke opened the pull request:
There is the case where OCs IDs might become outdated such as it has been with the calendar and contacts app which refer to the old dummy entry. This means that users with the old OCS id can't update updates as well will receive invalid state flags. (e.g. "experimental" instead of "approved")

To allow instances to properly update the applications in the future we need to migrate the OCS IDs for now manually.
# Testplan

Install one of the affected applications and run the repair steps and verify that the OCS id in the DB has been updated.
# Todo
- [x] Backport this PR to stable8.1 (https://github.com/owncloud/core/pull/17690)
- [x] Fix contacts OCS id (https://github.com/owncloud/contacts/pull/989 + https://github.com/owncloud/contacts/pull/990)
  - [x] Push update to appstore
- [x] Fix calendar OCS id (https://github.com/owncloud/calendar/pull/858 + https://github.com/owncloud/calendar/pull/859)
  - [x] Push update to appstore
- [x] Fix bookmarks OCS id (https://github.com/owncloud/bookmarks/pull/174 + https://github.com/owncloud/bookmarks/pull/175)
  - [x] Push update to appstore
- [x] Fix search_lucene OCS id (https://github.com/owncloud/search_lucene/pull/92 + https://github.com/owncloud/search_lucene/pull/93) 
  - [x] Push update to appstore
- [x] Fix documents OCS id
  - [x] Push update to appstore

<hr/>

@karlitschek As discussed.

LukasReschke committed:
Add repair step for outdated OCS IDs

There is the case where OCs IDs might become outdated such as it has been with the calendar and contacts app which refer to the old dummy entry. This means that users with the old OCS id can't update updates as well will receive invalid state flags. (e.g. "experimental" instead of "approved")

To allow instances to properly update the applications in the future we need to migrate the OCS IDs for now manually.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bcd84786-77f7-4f33-9294-2f6d01f30374) was created.

LukasReschke commented:
@jancborchardt This is wrt. to the issue you experienced. But notice that checking out the branch won't be enough as we also need to push the updates to the Appstore at one point then :speak_no_evil:

karlitschek commented:
Nice. As discussed!

LukasReschke commented:
Just noticed that we also need to prefer the ocsid from the DB instead the one in the info.xml in \OC_App::getAppInfo since another part of the update code relies on this behaviour. Will push later today.

LukasReschke committed:
Prefer OCS Id from database

To be consistent with other logic the app id from the database needs to be prefered. Especially when it comes to be able to replace an outdated OCS id.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6a5f2b77-367a-4131-a696-ca9df56ade7d) was created.

MorrisJobke commented:
@owncloud-bot retest this please

LukasReschke commented:
Good old borked Oracle. Gotta love this CI.

DeepDiver1975 commented:
@owncloud-bot retest this please

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
Jenkins PR #17719

MorrisJobke commented:
Jenkins is fine by with this: https://github.com/owncloud/core/pull/17719#issuecomment-122354816

nickvergessen commented on lib/repair/updateoutdatedocsids.php:
```diff
@@ -0,0 +1,108 @@
+	 * @param string $newId
+	 * @return bool True if updated, false otherwise
+	 */
+	public function fixOcsId($appName, $oldId, $newId) {
```
protected? no need to be public ;)

LukasReschke replied:
ACK

nickvergessen commented on lib/private/app.php:
```diff
@@ -659,6 +659,12 @@ public static function getAppInfo($appId, $path = false) {
+		if(isset($data['ocsid'])) {
+			$storedId = \OC::$server->getConfig()->getAppValue($appId, 'ocsid');
+			if($storedId !== '' && $storedId !== $data['ocsid']) {
+				$data['ocsid'] = $storedId;
```
So you prefer the DB value over the file value?

LukasReschke replied:
Yes.

LukasReschke replied:
Actually I'm not 100% sure if we need that though in 8.2 - I will trigger an update on the AppStore for some app and then I can check… :speak_no_evil:

MorrisJobke commented:
@LukasReschke Rebase?

MorrisJobke commented:
@LukasReschke Can we move this forward? :)

PVince81 commented:
After discussing with @LukasReschke on IRC, the repair step was put in 8.1 https://github.com/owncloud/core/pull/17690 and is not needed any more for versions coming after that, because skipping major versions is not allowed

PVince81 commented:
Hmm, but how about users who had 8.1.0 (without that repair step) and upgrade directly to 8.2 ?

Is this a direct forward port ?

LukasReschke commented:
> Is this a direct forward port ?

Direct cherry-pick.

LukasReschke commented:
Rebased.

PVince81 commented:
:+1:

Needs second reviewer @MorrisJobke @nickvergessen @georgehrke

DeepDiver1975 commented:
:+1:
